### PR TITLE
chore: build and export types [TCTC-3320]

### PR DESCRIPTION
**Pull request by ninofiliu**
# TL;DR

https://toucantoco.atlassian.net/browse/TCTC-3320
https://github.com/ToucanToco/tucana/pull/12553

Weaverbird is fully written in TS but didn't export type definitions, which frequently broke Tucana and other dependent packages. This PR adds a build script to generate type definitions to resolve this.

Before:

:x: no type defs, usual workaround in Tucana was to ts-ignore imports

image

:x: type errors not prevented

image

After:

:heavy_check_mark: type defs of imports

image

:heavy_check_mark: error prevention

image

# How

I first wanted to use rollup-plugin-dts, but it turned out that it required the latest rollup, the latest version of rollup plugins, the latest typescript, and after all these updates and many fixes in the code itself, the build was broken, likely due to a rollup error

```console
nino@dell~/tc/weaverbirdchore/dts$ yarn run build
yarn run v1.22.19
$ yarn build-bundle
$ rollup -c
clean: postcss.plugin was deprecated. Migration guide:
https://evilmartians.com/chronicles/postcss-8-plugin-migration

src/main.ts → dist/weaverbird.common.js, dist/weaverbird.esm.js, dist/weaverbird.browser.js...
(!) Plugin commonjs: The namedExports option from "@rollup/plugin-commonjs" is deprecated. Named exports are now handled automatically.
(!) Plugin replace: @rollup/plugin-replace: 'preventAssignment' currently defaults to false. It is recommended to set this option to `true`, as the next major version will default this option to `true`.
[!] Error: Unexpected character '@' (Note that you need plugins to import files that are not JavaScript)
src/components/Pagination.vue?rollup-plugin-vue=script.ts (29:0)
27: import { numberOfPages, pageMinMax, PaginationContext } from '@/lib/dataset/pagination';
28: 
29: @Component({
    ^
30:   name: 'pagination',
31:   components: {
Error: Unexpected character '@' (Note that you need plugins to import files that are not JavaScript)
    at error (/home/nino/tc/weaverbird/node_modules/rollup/dist/shared/rollup.js:198:30)
    at Module.error (/home/nino/tc/weaverbird/node_modules/rollup/dist/shared/rollup.js:12559:16)
    at Module.tryParse (/home/nino/tc/weaverbird/node_modules/rollup/dist/shared/rollup.js:12936:25)
    at Module.setSource (/home/nino/tc/weaverbird/node_modules/rollup/dist/shared/rollup.js:12841:24)
    at ModuleLoader.addModuleSource (/home/nino/tc/weaverbird/node_modules/rollup/dist/shared/rollup.js:22283:20)

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

I tried to fix this error but it looked like a messy thing to fix because things were supposed to work already: this is rollup failing to transpile a typescript file with decorators, eventhough decorator transpilation worked before all the updates, and went until page 5 of google to find answers and none worked.

I also tried to update plugin versions not to their latest, but I either got the error above or heap overflow errors. At this point I abandonned.

All wasn't lost tho, the build broke only on the last commit, so there's this free work available if we want weaverbird to be built with the latest typescript and rollup version (but not with the latest plugin versions apparently).

The branch is [chore/dts](https://github.com/ToucanToco/weaverbird/compare/06dc57800d703a0113afeca3d2b6e28eced2876d...chore/dts) if you wanna take a look.

# Why the build script? Why not use TSC directly to generate types?

If we generate types they'll still have path aliases in them. Stuff that looks like:

```ts
// /dist/types/main.d.ts
import foo from '@/bar/baz/foo';
```

and these path won't resolve when being used in dependent modules. Everybody wants TSC to natively resolve paths aliases when building declaration files https://github.com/microsoft/TypeScript/issues/15479 https://github.com/microsoft/TypeScript/issues/5039 but it has been refused several times so I figured that having a dirty but simple sed-based script to replace '@' by the path to src in declaration files was the simplest workaround.

# Should we worry about the impact of this?

Probably not.

Build time time only lasts 7% longer.

```console
nino@dell~/tc/weaverbirdmaster$ time npx rollup -c
[...]
real	0m56.346s
user	1m25.636s
sys	0m4.164s
nino@dell~/tc/weaverbirdchore/dts2$ time . build-types.sh 
real    0m3.966s
user	0m7.545s
sys	0m0.278s
```

The only thing I'd probably worry a tiny bit about would be the use of the [dot command](https://en.wikipedia.org/wiki/Dot_(command)) but it's a POSIX standard so should work on mac on unix, and also [works in powershell](https://devblogs.microsoft.com/powershell/powershell-constrained-language-mode-and-the-dot-source-operator/#:~:text=The%20PowerShell%20dot%2Dsource%20operator,file%20directly%20into%20your%20script).

### CODE REVIEW

**Comment by ninofiliu** [comment]
ah shit i forgot about tests i'm on it yo

**Comment by ninofiliu** [comment]
CI fails with

```
55: export { PipelineStep, Pipeline } from './lib/steps';
             ^
Error: 'PipelineStep' is not exported by src/lib/steps.ts
```

but that's baloney, I don't know what's happening, I'm rebasing rn

**Comment by ninofiliu** [comment]
Oooooh okay that was a hard one

[The docs](https://rollupjs.org/guide/en/#error-name-is-not-exported-by-module) say that it comes from the commonjs plugin not inferring the right named exports, so I tried every combination of

```ts
// rollup.config.js
export default {
  // ...
  plugins: [
    // ...
    commonjs({
      namedExports: {
        'src/lib/steps.ts': ['PipelineStep', 'Pipeline'],
        // ...
      },
    });
 ]
};
```

with and without the ts, with relative path, with full path, with the path as it appears in the file, etc, but to no avail. I upgraded the plugin to latest also, still the same error. After a while I gave up because I didn't expect that much from a deprecated plugin.

Instead I decided to have a specific file for the type defs `src/types.ts` so that the main rollup pipeline should not even worry about these types, so now that should be good

**Comment by ninofiliu** [comment]
OH MY GOD SONAR CAN YOU BE NOT STUPID FOR LIKE HALF A SECOND AND NOT CHECK THE CODE COVERAGE OF A FILE THAT JUST PASSES EXPORTS AROUND

**Comment by ninofiliu** [comment]
FINALLY the CI is green, god that was exhausting

image

File: build-types.sh [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#!/bin/sh
+rm -rf dist/types
+npx tsc -p tsconfig.types.json
```
[the remark this answers is not in the document]
Comment (ninofiliu) [reply build-types.sh:3]: Wait, isn't npx available wherever node is? If not I don't have any preferences. Maybe the dot command erases the path? Not sure. Probably the safer thing would be to replace `npx tsc` by `node node_modules/.bin/tsc` then?

File: build-types.sh [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+for file in dist/types/**/*.*
+do
+  if [ $file = dist/types/main.d.ts ]
+  then
+    path_to_src=./
+  else
+    path_to_src=$(echo $file | tr -dc '/' | colrm 1 2 | sed "s#/#../#g")
+  fi
+  sed -i -r "s#((import|export) .* from ')@/(.*)#\1$path_to_src\3#g" $file
+done
```
[the remark this answers is not in the document]
Comment (ninofiliu) [reply build-types.sh:13]: Didn't know about this one! Yeah let's switch to AST-based codemods if this build pipeline get too complex

File: build-types.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (ninofiliu) [reply build-types.sh:3]: Okay so apparently [npx is available wherever npm is](https://nodejs.dev/en/learn/the-npx-nodejs-package-runner) and I doubt that it's necessary to handle build systems where npm is not available

[the remark this answers is not in the document]
Comment (davinov) [reply build-types.sh:3]: Oh nice! I thought it was standalone (https://www.npmjs.com/package/npx)
Good to use npx then!
